### PR TITLE
Issue9 synchronized sink

### PR DESCRIPTION
**Pull request by DannyPeck**
See: https://github.com/CachedNerds/Toolbox/wiki/Thread-Safe-Sink-Design for more design information.

### CODE REVIEW

File: C++/Toolbox/Sink/BasicSink.h [hunk 1]
```diff
@@ -3,23 +3,36 @@
...
-  BasicSink<OutputType> & operator << (const OutputType & output)
+  using value_type = OutputType;
+
+  BasicSink (void) = default;
+  BasicSink (const BasicSink & rhs) = delete;
+  BasicSink & operator= (const BasicSink & rhs) = delete;
+  BasicSink (BasicSink && rhs) = default;
+  BasicSink & operator= (BasicSink && rhs) = default;
+
+  ~BasicSink (void)
```
Comment (DannyPeck) [inline C++/Toolbox/Sink/BasicSink.h]: This needs to be virtual.

**Comment by xmclark** [comment]
I want to be able to comment on your design page. Damn you confluence for making me expect certain features!

File: C++/Toolbox/Sink/ThreadSafeProxySink.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+}
+
+ThreadSafeProxySink (ThreadSafeProxySink && sink)
+: _sink (std::move (sink._sink))
+{
+
+}
+
+ThreadSafeProxySink & operator= (ThreadSafeProxySink && sink)
```
Comment (xmclark) [inline C++/Toolbox/Sink/ThreadSafeProxySink.h:31]: Regarding both the copy constructor and copy assignment operator: I'm not sure of the best practices here, but does it seem more natural to you @DannyPeck to leverage the `default` and `delete` keywords here? You are effectively converting your copy constructor into a move constructor and your copy assignment operator into a move assignment operator. Why not simply `delete` the copy constructor/assignment and either define or `default` the move constructor/assignment?

File: C++/libs/valgrind.txt [hunk 3]
```diff
@@ -0,0 +1,172 @@
+==36284== Memcheck, a memory error detector
```
Comment (xmclark) [inline C++/libs/valgrind.txt]: Why is this file here? Is this configuration for valgrind?

File: C++/libs/Tuprules.tup [hunk 4]
```diff
@@ -3,7 +3,7 @@
 CC = g++
 
 # Uncomment to build with debug symbols
-#CFLAGS += -g
+CFLAGS += -g
```
Comment (xmclark) [inline C++/libs/Tuprules.tup]: This doesn't really bother me, but I'm hoping [Hatch](https://github.com/CachedNerds/Hatch) will help us with managing debug/release builds better.

File: C++/libs/Sink/test/src/ThreadSafeSink.cpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  std::string message = "test";
+  proxySink << message;
+  
+  REQUIRE (rc == message);
+}
+
+TEST_CASE ("Instantiate ThreadSafeProxySink using make_thread_safe")
+{
+  std::string rc;
+  ThreadSafeProxySink<Test::Sink> proxySink = make_thread_safe<Test::Sink>(rc);
```
Comment (xmclark) [inline C++/libs/Sink/test/src/ThreadSafeSink.cpp]: I like this.

Comment (xmclark) [reply C++/libs/Sink/test/src/ThreadSafeSink.cpp]: when I write code using ThreadSafeSink, I will opt for auto instead:
```
auto sink = make_thread_safe_sink<Test::Sink>(rc);
```

**Review by xmclark — APPROVED** [review]

File: C++/Toolbox/Sink/ThreadSafeProxySink.h (hunk 2 above)
Comment (DannyPeck) [reply C++/Toolbox/Sink/ThreadSafeProxySink.h:31]: That is actually the Move Constructor and Move Assignment Operator.  And interestingly, if I switch to using the default key word for those methods it tries to use the copy constructor and it doesn't work.  Also the copy constructors are explicitly deleted in the BasicSink base class so they are implicitly deleted in all subclasses.  So in conclusion, that is exactly what I did.  Is my signature for the move constructor wrong?

File: C++/libs/Sink/test/src/ThreadSafeSink.cpp (hunk 5 above)
Comment (DannyPeck) [reply C++/libs/Sink/test/src/ThreadSafeSink.cpp]: I agree.

File: C++/libs/Tuprules.tup (hunk 4 above)
Comment (DannyPeck) [reply C++/libs/Tuprules.tup]: Thank you for noticing this, this was simply something I forgot to disable after a debugging session.  I agree completely about Hatch.

File: C++/libs/valgrind.txt (hunk 3 above)
Comment (DannyPeck) [reply C++/libs/valgrind.txt]: That was also an accidental artifact left over from a debugging session.  It has been removed.

File: C++/libs/Sink/test/src/ThreadSafeSink.cpp (hunk 5 above)
Comment (DannyPeck) [reply C++/libs/Sink/test/src/ThreadSafeSink.cpp]: Also, I have named the function `make_thread_safe` instead of `make_thread_safe_sink` following the same way the smart pointer make functions work, `make_unique` not `make_unique_ptr`

File: C++/Toolbox/Sink/ThreadSafeProxySink.h (hunk 2 above)
Comment (xmclark) [reply C++/Toolbox/Sink/ThreadSafeProxySink.h:31]: Yeah now that I look at this...I think I was just really tired. You're totally correct. This is the move constructor/assignment operator. Your syntax is correct. My bad 😴

**Review by xmclark — APPROVED** [review]

**Review by xmclark — APPROVED** [review]

**Review by technetos — APPROVED** [review]

File: C++/Toolbox/Sink/ThreadSafeProxySink.h (hunk 2 above)
Comment (DannyPeck) [reply C++/Toolbox/Sink/ThreadSafeProxySink.h:31]: You're good!  I still think it's weird that I had to define it and that default didn't work...
